satpy/readers/li_l2_nc.py

```diff
@@ -158,11 +158,9 @@ def get_array_on_fci_grid(self, data_array: xr.DataArray):
         data_2d = da.map_blocks(_np_add_at_wrapper, data_2d, (rows, cols), data_array,
                                 dtype=data_array.dtype,
                                 chunks=(LI_GRID_SHAPE[0], LI_GRID_SHAPE[1]))
-        data_2d = da.where(data_2d > 0, data_2d, np.nan)
-
+        data_2d = da.where(data_2d > 0, data_2d, np.nan).astype(np.float32)

```

Comment:
This would prevent one unnecessary upcasting:
```suggestion
        data_2d = da.where(data_2d > 0, data_2d, np.float32(np.nan))

```

Reply:
@pnuu  casting a `nan` who is a float32 is not enough to convert all the arrays into a `float32` . For example if `data_2d` is an `int32` the method where will convert it to a `float64`. To prevent it I have used the method `astype(np.float32)`.

Reply:
Ok, that makes sense I guess.

But for integer data the data would still be converted to floats, which seemed to be the original problem reported https://github.com/pytroll/satpy/issues/2854 , right? So should there be separate handling for the integer data?

Reply:
For this specific case, handling the accumulated LI 2-d data arrays, we need floats since we need to support NaN values. So I'm ok with the solution here that avoids float64.

The problem in the original issue was that some integer variables, that do not have a FillValue attribute, were still being casted unnecessarily to float; that problem is fixed by the other modification of this PR in `apply_fill_value` here https://github.com/pytroll/satpy/pull/2998/files#diff-3b2bff08b4001ec6f72cca67791cc4322b38e0db97e68f2109791093e56e6052R445

Reply:
Oh, ok.

The current version will cast the data to `float64` (when setting the 64-bit `nan`) before recasting it to `float32`. So for the staying in `float32` world I think the `data_2d` array should be converted to `float32` before applying the `nan`:

```python
data_2d = data_2d.astype(np.float32)
data_2d = da.where(data_2d > 0, data_2d, np.nan)
```

Here the `np.nan` respects the original `dtype` and does not do an intermediate up-casting for the integer data.

Reply:
Modification applied